docs: libp2p resolver rfc (#629)
eigerco/polka-storage: rfcs/001-libp2p-resolver.md

# RFC-001: Peer ID to Multi-Address Resolver

> 

Reviewer: You're missing the author + date

# RFC-001: Peer ID to Multi-Address Resolver

## Abstract

> This RFC proposes an application for resolving [Peer ID][5]s to [multi-addresses][4] in a [libp2p][1] network.

Reviewer 2: Why do we need a separate application for that? Shouldn't it be a part of file retrieval or something similar?

Author: IMO, the file retrieval interface should be a bit simpler for the end-user, they don't care about peer ID resolver if they already have the multiaddr, they just want to retrieve their file with the multiaddr and the CID. This application should be written so the retriever can use the functionality to resolve the multiaddr from the peer ID.

Reviewer 2: So we're proposing a `library` or an standalone application for resolving PeerIDs?
So this libp2p resolver will be a separate system process?

The application will utilize [libp2p][1]'s [rendezvous protocol][2] for peer discovery and the [identify protocol][3] for retrieving [multi-addresses][4] and related information.
By leveraging these protocols, the application aims to provide an efficient and seamless mechanism for peer-to-peer address resolution over the internet.

## Introduction

In [libp2p][1]-based networks, [Peer ID][5]s serve as unique identifiers for peers, while [multi-addresses][4] describe their network locations.

> Clients need to resolve [Peer ID][5]s to [multi-addresses][4] do they can connect to storage providers and fetch their files. File fetching will not be covered in this document.

Reviewer: More general + more accurate (since you can send a status request)
```suggestion
Clients need to resolve [Peer ID][5]s to [multi-addresses][4] so they can connect to and communicate with storage providers.
```

This application will provide a straightforward interface to resolve [Peer ID][5]s to [multi-addresses][4] by using [libp2p][1]'s discovery and information exchange mechanisms.

The application will operate over the internet, relying on bootstrap nodes as rendezvous points to join the network and initiate discovery.
It will integrate [libp2p][1]'s [rendezvous protocol][2] for locating peers and the [identify protocol][3] for exchanging relevant connection information.

## Bootstrap Nodes as Rendezvous Points

The application will rely on bootstrap nodes to initialize network interactions.
These nodes act as rendezvous points, enabling the clients to join the network and locate other peers.
Clients can configure the bootstrap nodes through a configuration file or command-line arguments.

### Bootstrap Node Incentives

> Storage providers have a incentive to run bootstrap nodes in the P2P network because it enhances their trustworthiness and visibility within the ecosystem.

Reviewer: Running a bootstrap node doesn't make you more trustworthy, it's just perception. Correctly handling deals makes you trustworthy!
```suggestion
Storage providers have a incentive to run bootstrap nodes in the P2P network because it increases their perceived trustworthiness and visibility within the ecosystem.
```

As bootstrap nodes, storage providers aid in peer discovery, enabling new and existing peers to connect more easily to the network.
This presence not only benefits the network's overall health and connectivity but also reflects positively on the storage provider's reliability and commitment to the decentralized storage ecosystem.

> Trust is critical factor for storage clients selecting which storage provider to store their data.

Reviewer: Note: we don't have a trust/reputation mechanism, as such we need to be careful with our wording. I'm not sure that removing this line would make it better but I'm also not sure it should be kept as is
```suggestion
Trust is a critical factor for storage clients selecting which storage provider to store their data.
```

A provider that operates a bootstrap node demonstrates that they want to support the broader network, increasing confidence and trustworthiness among clients.

> Bootstrap nodes and contribute to a more seamless client experience by helping ensure that content can be discovered and retrieved efficiently.

Reviewer: ```suggestion
Bootstrap nodes are required for a seamless client experience by helping ensure that content can be discovered and retrieved efficiently.
```

- Initiate the [identify protocol][3] handshake to request identity information.
- Extract and display the [multi-addresses][4] and associated data provided by the target peer.

The application will output this information in a human-readable format, making it easy for clients to consume and utilize in further operations.
The combination of the [rendezvous protocol][2] and [identify protocol][3] ensures a robust and modular approach to both discovering and resolving peers in the network.

## Command-Line Interface

The application will provide an intuitive CLI interface, allowing clients to input a [Peer ID][5] and retrieve its [multi-addresses][4].
Additional options will enable customization of settings, such as specifying bootstrap nodes or adjusting discovery parameters.
The output will include a list of [multi-addresses][4] associated with the [Peer ID][5], presented in a structured and readable format.

## Configuration and Deployment

The application will allow configuration through a settings file or environment variables.
Clients can define bootstrap nodes, set timeouts for peer discovery and identify protocol exchanges, and specify logging preferences.

> The CLI tool will be distributed as a single executable binary, ensuring ease of deployment across diverse platforms and environments.

Reviewer: I don't think this is relevant, IMO the RFC is to discuss the high-level ideas; CLI, single or multiple binaries, etc, those are implementation details.

https://github.com/eigerco/polka-storage/pull/629/files#diff-96ebdac0b36633969fe397b19ef41c994f26b798a7bff824a8ca1f5486aa0f7eR22

This line can almost be thought of as the border, it states that the final product should be configurable, provides and example, but it doesn't really commit to anything
